src/core/model.jl

```diff
@@ -84,21 +105,28 @@ warnings are thrown when appropriate.
 function AgentBasedModel(
     ::Type{A},
     space::S = nothing;
+    container::Type = Dict{Int,A},
     scheduler::F = Schedulers.fastest,
     properties::P = nothing,
     rng::R = Random.default_rng(),
-    warn = true,
+    warn = true
 ) where {A<:AbstractAgent,S<:SpaceType,F,P,R<:AbstractRNG}
     agent_validator(A, space, warn)
-
-    agents = Dict{Int,A}()
-    return ABM{S,A,F,P,R}(agents, space, scheduler, properties, rng, Ref(0))
+    C = construct_agent_container(container, A)
+    agents = C()
+    return ABM{S,A,C,F,P,R}(agents, space, scheduler, properties, rng, Ref(0))
 end
 
 function AgentBasedModel(agent::AbstractAgent, args...; kwargs...)
     return ABM(typeof(agent), args...; kwargs...)
 end
 
+"""
+    UnkillableABM(AgentType [, space]; properties, kwargs...) → model
+Create an unkillable agent-based model from the given agent type and `space`. For more details see `ABM`.
```

Comment:
Let's say a bit more here. I propose:
"""
Same as [`ABM`](@ref), but agents cannot be removed during model stepping. Additionally, it is mandatory that the agent ID is exactly the same as the agent insertion order (i.e., the 5th agent added to the model **must** have ID 5). This allows the internal storage of agents to be in a standard `Vector`, accelerating a lot the performance of retrieving agents from the model.
"""

This of course means that in the function `add_agent!`  you must add a dispatch method for `UnkillableABM` which enforces that an added agent must have ID that is the amount of current agents + 1.

Reply:
Done with d2dd559082a971ac8ea066c459f9a514f2a121e5.